src/blockdata/witness.rs

```diff
@@ -0,0 +1,327 @@
+    }
+}
+
+impl Witness {
+    /// Returns `true` if the witness contains no element
+    pub fn is_empty(&self) -> bool {
+        self.witness_elements == 0
+    }
+
+    /// Returns a struct implementing [`Iterator`]
+    pub fn iter(&self) -> WitnessIterator {
+        WitnessIterator {
+            witness: &self,
+            cursor: 0,
+        }
+    }
+
+    /// Returns the number of elements this witness holds
+    pub fn len(&self) -> usize {
+        self.witness_elements as usize
+    }
+
+    /// Returns the bytes required when this Witness is consensus encoded
+    pub fn serialized_len(&self) -> usize {
+        self.iter()
+            .map(|el| VarInt(el.len() as u64).len() + el.len())
+            .sum::<usize>()
+            + VarInt(self.witness_elements).len()
+    }
+
+    /// Clear the witness
+    pub fn clear(&mut self) {
+        self.content.clear();
+        self.witness_elements = 0;
+    }
+
+    /// Push a new element on the witness, require an allocation
+    pub fn push(&mut self, new_element: &[u8]) {
+        self.witness_elements += 1;
+        let element_len_varint = VarInt(new_element.len() as u64);
+        let current_content_len = self.content.len();
+        self.content.resize(
+            current_content_len + element_len_varint.len() + new_element.len(),
+            0,
+        );
+        let end_varint = current_content_len + element_len_varint.len();
+        element_len_varint
+            .consensus_encode(&mut self.content[current_content_len..end_varint])
+            .expect("writers on vec don't error, space granted through previous resize");
+        self.content[end_varint..].copy_from_slice(new_element);
+    }
+}
+
+impl Default for Witness {
+    fn default() -> Self {
+        // from https://doc.rust-lang.org/std/vec/struct.Vec.html#method.new
+        // The vector will not allocate until elements are pushed onto it.
+        Witness {
+            content: Vec::new(),
+            witness_elements: 0,
+        }
+    }
+}
+
+impl<'a> Iterator for WitnessIterator<'a> {
+    type Item = &'a [u8];
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let vec = &self.witness.content;
+        if self.cursor >= vec.len() {
+            None
+        } else {
+            let var = VarInt::consensus_decode(&vec[self.cursor..])
+                .expect("is granted witness.content contains varint because created only from internal methods");
+            let start = self.cursor + var.len();
+            let end = start + var.0 as usize;
+            self.cursor = end;
+            Some(&vec[start..end])
+        }
+    }
+}
+
+// Serde keep backward compatibility with old Vec<Vec<u8>> format
+#[cfg(feature = "serde")]
+impl serde::Serialize for Witness {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let vec: Vec<_> = self.iter().map(|e| e.to_vec()).collect();
```

Comment:
I think this could be faster but for initial code probably good. (Manual serde impls are annoying as hell.)